File: matlab/+smalltargetmotiondetectors/+core/+apgstmd_core/PredictionModule.m
classdef PredictionModule < smalltargetmotiondetectors.core.BaseCore
    % PredictionModule class for ApgSTMD

    properties
        % Parameters
        velocity;           % Velocity: v_{opt} (Defalt: 0.25)
        intDeltaT = 25;     % Delta time
        sizeFilter = 25;    % Size of filter
        numFilter = 8;      % Number of filters
        zeta = 2;           % Zeta parameter
        eta = 2.5;          % Eta parameter
        kappa = 0.02;       % Kappa parameter
        mu = 0.75;          % Mu parameter
        beta = 1;           % Beta parameter
    end

    properties(Hidden)
        % Hidden properties
        predictionKernel;       % Prediction kernel
        cellPredictionGain;     % Cell array for prediction gain
        cellPredictionMap;      % Cell array for prediction map
        timeAttenuationKernel;  % kernel in formula (23)
    end

    methods
        function self = PredictionModule()
            % Constructor method
            % Initializes the PredictionModule object
            
            self = self@smalltargetmotiondetectors.core.BaseCore();
        end
    end

    methods
        function init(self)
            % Initialization method
            % Initializes the prediction module
            
            import smalltargetmotiondetectors.tool.kernel.*;

            if self.intDeltaT < 1
                self.intDeltaT = 1;
            elseif ~isinteger(self.intDeltaT)
                self.intDeltaT = round(self.intDeltaT);
            end
            
            if isempty(self.velocity)
                self.velocity = 25 / 4 / self.intDeltaT;
            end
            
            self.predictionKernel = create_prediction_kernel(...
                self.velocity, ...
                self.intDeltaT, ...
                self.sizeFilter, ...
                self.numFilter, ...
                self.zeta, ...
                self.eta);
            self.cellPredictionGain = ...
                cell(self.intDeltaT + 1, self.numFilter);
            self.cellPredictionMap = cell(self.intDeltaT + 1, 1);

            self.timeAttenuationKernel = ...
                exp( self.kappa * ( -self.intDeltaT:0 ) );
            
        end

        function [facilitatedOpt, predictionMap] = process(self, lobulaOpt)
            % Processing method
            % Processes the input lobulaOpt to predict motion and update
            % prediction map
            import smalltargetmotiondetectors.tool.compute.*;
            
            numDict = length(lobulaOpt);
            [imgH, imgW] = size(lobulaOpt{1});

            % Prediction Gain
            self.cellPredictionGain = circshift(self.cellPredictionGain, -1);
            predictionGain = cell(numDict, 1);
            for idxD = 1:numDict
                if isempty(self.cellPredictionGain{1,idxD})
                    predictionGain{idxD} = conv2(...
                        self.mu * lobulaOpt{idxD},...
                        self.predictionKernel{idxD},...
                        'same' ...
                        );
                else
                    predictionGain{idxD} = conv2(...
                        self.mu * lobulaOpt{idxD} ...
                        + (1 - self.mu) * self.cellPredictionGain{1,idxD}, ...
                        self.predictionKernel{idxD}, ...
                        'same' ...
                        );
                end
            end

            [self.cellPredictionGain(end,:)] = deal(predictionGain);

            % Prediction Map
            tobePredictionMap = zeros(imgH, imgW);
            for idxD = 1:numDict
                tobePredictionMap = tobePredictionMap + predictionGain{idxD};
            end


            % Facilitated STMD Output
            facilitatedOpt = cell(numDict, 1);
            for idxD = 1:numDict
                facilitatedOpt{idxD} = lobulaOpt{idxD};
                if ~isempty(self.cellPredictionGain{end, idxD})
                    facilitatedOpt{idxD} = facilitatedOpt{idxD} + ...
                        self.beta * ...
                        compute_temporal_conv( ...
                        self.cellPredictionGain(:,idxD), ...
                        self.timeAttenuationKernel ...
                        );
                end

            end

            % Memorizer update
            maxTobePreMap = max(tobePredictionMap, [], 'all');
            self.cellPredictionMap = circshift(self.cellPredictionMap, -1);
            self.cellPredictionMap{end} = ...
                ( tobePredictionMap > maxTobePreMap*2e-1 );

            % Output
            predictionMap = self.cellPredictionMap{1};
            self.Opt = facilitatedOpt;
        end
    end

end

File: python/smalltargetmotiondetectors/core/apgstmd_core/PredictionModule.py
import numpy as np
from smalltargetmotiondetectors.core.base_core import BaseCore
from smalltargetmotiondetectors.util.kernel import create_prediction_kernel
from smalltargetmotiondetectors.util.compute import compute_temporal_conv

class PredictionModule(BaseCore):
    """
    PredictionModule class for ApgSTMD.
    
    This class implements the prediction module in the ApgSTMD.
    """
    
    def __init__(self):
        """
        Constructor method.
        
        Initializes the PredictionModule object.
        """
        super().__init__()
        self.velocity = None
        self.intDeltaT = 25
        self.sizeFilter = 25
        self.numFilter = 8
        self.zeta = 2
        self.eta = 2.5
        self.kappa = 0.02
        self.mu = 0.75
        self.beta = 1
        self.prediction_kernel = None
        self.cell_prediction_gain = None
        self.cell_prediction_map = None
        self.time_attenuation_kernel = None
    
    def init_config(self):
        """
        initiate config for prediction module.
        """
        from smalltargetmotiondetectors.util.kernel import create_prediction_kernel
        
        if self.intDeltaT < 0:
            self.intDeltaT = 0
        elif not isinstance(self.intDeltaT, int):
            self.intDeltaT = round(self.intDeltaT)
        
        if self.velocity is None:
            self.velocity = 25 / 4 / self.intDeltaT
        
        self.prediction_kernel = create_prediction_kernel(
            self.velocity,
            self.intDeltaT,
            self.sizeFilter,
            self.numFilter,
            self.zeta,
            self.eta
        )
        self.cell_prediction_gain = [[None]*self.numFilter for _ in range(self.intDeltaT + 1)]
        self.cell_prediction_map = [None]*(self.intDeltaT + 1)
        
        self.time_attenuation_kernel = np.exp(self.kappa * np.arange(-self.intDeltaT, 1))
    
    def process(self, lobula_opt):
        """
        Processing method.
        
        Processes the input lobula_opt to predict motion and update
        prediction map.
        """
        num_dict = len(lobula_opt)
        img_h, img_w = lobula_opt[0].shape
        
        # Prediction Gain
        self.cell_prediction_gain = np.roll(self.cell_prediction_gain, shift=-1, axis=0)
        prediction_gain = []
        for idxD in range(num_dict):
            if self.cell_prediction_gain[0][idxD] is None:
                prediction_gain.append(np.convolve(
                    self.mu * lobula_opt[idxD],
                    self.prediction_kernel[idxD],
                    mode='same'
                ))
            else:
                prediction_gain.append(np.convolve(
                    self.mu * lobula_opt[idxD] + (1 - self.mu) * self.cell_prediction_gain[0][idxD],
                    self.prediction_kernel[idxD],
                    mode='same'
                ))
        self.cell_prediction_gain[-1] = prediction_gain
        
        # Prediction Map
        tobe_prediction_map = np.zeros((img_h, img_w))
        for idxD in range(num_dict):
            tobe_prediction_map += prediction_gain[idxD]
        
        # Facilitated STMD Output
        facilitated_opt = [np.copy(lobula_opt[idxD]) for idxD in range(num_dict)]
        for idxD in range(num_dict):
            if self.cell_prediction_gain[-1][idxD] is not None:
                facilitated_opt[idxD] += self.beta * compute_temporal_conv(
                    self.cell_prediction_gain[:, idxD],
                    self.time_attenuation_kernel
                )
        
        # Memorizer update
        max_tobe_pre_map = np.max(tobe_prediction_map)
        self.cell_prediction_map = np.roll(self.cell_prediction_map, shift=-1)
        self.cell_prediction_map[-1] = (tobe_prediction_map > max_tobe_pre_map * 2e-1)
        
        # Output
        prediction_map = self.cell_prediction_map[0]
        self.Opt = facilitated_opt
        return facilitated_opt, prediction_map
